build/lib/src/asset/id.dart

```diff
@@ -22,13 +22,57 @@ class AssetId implements Comparable<AssetId> {
   /// Gets the file extension of the asset, if it has one, including the ".".
   String get extension => pathos.extension(path);
 
-  /// Creates a new AssetId at [path] within [package].
+  /// Creates a new [AssetId] at [path] within [package].
   ///
   /// The [path] will be normalized: any backslashes will be replaced with
   /// forward slashes (regardless of host OS) and "." and ".." will be removed
   /// where possible.
   AssetId(this.package, String path) : path = _normalizePath(path);
 
+  /// Creates a new [AssetId] from an [uri].
+  ///
+  /// This gracefully handles `package:` URIs, which is useful when creating an
+  /// [AssetId] from an `import` or `export` directive pointing to a package:
+  /// ```dart
+  /// AssetId assetOfDirective(UriReferencedElement element) {
+  ///   return new AssetId.resolve(element.uri);
```

Comment:
So `element.librarySource.source` used to be an `AssetBasedSource`, which had an `id` field. However I think with the summaries resolver that may not be the case which makes things a bit more complicated. Would you just file an issue to support more than `package:` URIs and we can follow up.